### DISCUSSION TITLE
x-transition without x-show

### DISCUSSION THREAD

**Discussion by giorgiopogliani** [alpinejs/alpine#947] [1. Help]
Hi, 

is it possible to use x-transition:* without x-show? I think it makes sense not to hide always the thing or I am missing something?

codepen: https://codepen.io/giorgiopogliani/pen/WNGEyXd

```html
<div x-data="{ toggle: false }" class="min-h-screen flex items-center justify-center">
  <button 
    @click.prevent="toggle = !toggle" 
    class="px-4 h-10 flex items-center justify-center bg-gray-50 border shadow rounded"
  >
    <!-- What if I don't to hide the thing --
    <!-- but I just want to animate the transition? -->
    <span
      class="h-4 w-4 block"
      x-show="toggle"
      x-transition:enter="bg-gray-200 transition ease-out duration-300"
      x-transition:enter-start="bg-blue-300 transform scale-90"
      x-transition:enter-end="bg-blue-400 transform scale-100"
      x-transition:leave="transition ease-in duration-300"
      x-transition:leave-start="bg-blue-500 transform scale-100"
      x-transition:leave-end="bg-blue-600 transform scale-90"  
    >
    </span>
    Test
  </button>
</div>
```

**Comment by KevinBatdorf** [comment]
You would do something like this:

```html
<span
  class="h-4 w-4 block bg-blue-300 transform transition ease-out duration-300"
  :class="{'scale-50': toggle}"
>
```

**Reply by giorgiopogliani** [reply]
Well, yes, but it's not the really the same api.

**Reply by KevinBatdorf** [reply]
Yeah, that's tied to `x-show`/`x-if`. You could probably create your own similar api via `$watch`/`$nextTick` to then remove some classes if you need more control.

**Reply by SimoTod** [reply]
You need something to trigger the change though. If you just put a x-transition to a random element, alpine won't know when it should start the transition right? The transition api only works when you show/hide an element. for other custom use case, the simplest way is to use :class as Kevin suggested

**Comment by robinp** [comment]
The need to trigger independently of `x-show` is especially pronounced with the Intersect plugin's `x-intersect`. Having to hide an element to be flown-in smoothly will lead to other UI elements jumping around when the element becomes visible.

Note: https://github.com/alpinejs/alpine/issues/361 is slightly related.

Two options I see (no idea about Alpine.js internals):
- make it possible to set which display state is used by `x-show`. In this case `hidden` would be more appropriate, as then one could pre-reserve the space for the element, just not show it (at least for horizontal slide-in that would help)
- make it customizable for `x-transition` which variable would it be toggled by. Thet `x-intersect` could just set that instead `shown`.

Of course, I could just shut up and code. But as a hint for future reference.

**Reply by ekwoka** [reply]
x-transition doesn't really make much sense to use for anything other than x-show/x-if.

A normal state with `:class` does the whole thing pretty much.

**Reply by robinp** [reply]
@ekwoka thank you, that's a nice idea! For the record, this is the diff that did the trick (for a horizontal slide + fade-in):

```
  <div x-data="{ shown: false }" x-intersect.once="shown = true">
-    <div x-cloak x-show="shown"
-      x-transition:enter="transition duration-1000"
-      x-transition:enter-start="translate-x-3/4 opacity-0"
-      x-transition:enter-end="translate-x-none opacity-1"
+    <div
+      :class="shown ? 'transition translate-x-none opacity-1 duration-1000' : 'translate-x-3/4 opacity-0 duration-0'"
       >
```

**Reply by kirkbushell** [reply]
@robinp I did that initially, but found other issues with taking that approach - namely, if you're scrolled down the page on a refresh, or going back, it resulted in a lot of weirdness.

Your suggestion above in the first response is really what we need - x-intersect currently is very limited, and results in some really bizarre outcomes. As you said, it really only appears to have been thought of within the context of show/if.

For me, all I want to do is apply classes for doing things like sliding in elements or text as you scroll, and it's ended up being a real headache.
